recipes/faiss/meta.yaml

```diff
@@ -0,0 +1,204 @@
+    'gpu/impl/IVFAppend.cuh', 'gpu/impl/IVFBase.cuh', 'gpu/impl/IVFFlat.cuh',
+    'gpu/impl/IVFFlatScan.cuh', 'gpu/impl/IVFPQ.cuh', 'gpu/impl/IVFUtils.cuh',
+    'gpu/impl/L2Norm.cuh', 'gpu/impl/L2Select.cuh', 'gpu/impl/Metrics.cuh',
+    'gpu/impl/PQCodeDistances-inl.cuh', 'gpu/impl/PQCodeDistances.cuh', 'gpu/impl/PQCodeLoad.cuh',
+    'gpu/impl/PQScanMultiPassNoPrecomputed-inl.cuh', 'gpu/impl/PQScanMultiPassNoPrecomputed.cuh',
+    'gpu/impl/PQScanMultiPassPrecomputed.cuh', 'gpu/impl/VectorResidual.cuh',
+    'gpu/utils/DeviceMemory.h', 'gpu/utils/DeviceUtils.h', 'gpu/utils/MemorySpace.h',
+    'gpu/utils/StackDeviceMemory.h', 'gpu/utils/StaticUtils.h', 'gpu/utils/Timer.h',
+    'gpu/utils/BlockSelectKernel.cuh', 'gpu/utils/Comparators.cuh',
+    'gpu/utils/ConversionOperators.cuh', 'gpu/utils/CopyUtils.cuh', 'gpu/utils/DeviceDefs.cuh',
+    'gpu/utils/DeviceTensor-inl.cuh', 'gpu/utils/DeviceTensor.cuh', 'gpu/utils/DeviceVector.cuh',
+    'gpu/utils/Float16.cuh', 'gpu/utils/HostTensor-inl.cuh', 'gpu/utils/HostTensor.cuh',
+    'gpu/utils/Limits.cuh', 'gpu/utils/LoadStoreOperators.cuh', 'gpu/utils/MathOperators.cuh',
+    'gpu/utils/MatrixMult-inl.cuh', 'gpu/utils/MatrixMult.cuh', 'gpu/utils/MergeNetworkBlock.cuh',
+    'gpu/utils/MergeNetworkUtils.cuh', 'gpu/utils/MergeNetworkWarp.cuh',
+    'gpu/utils/NoTypeTensor.cuh', 'gpu/utils/Pair.cuh', 'gpu/utils/PtxUtils.cuh',
+    'gpu/utils/ReductionOperators.cuh', 'gpu/utils/Reductions.cuh', 'gpu/utils/Select.cuh',
+    'gpu/utils/Tensor-inl.cuh', 'gpu/utils/Tensor.cuh', 'gpu/utils/ThrustAllocator.cuh',
+    'gpu/utils/Transpose.cuh', 'gpu/utils/WarpSelectKernel.cuh', 'gpu/utils/WarpShuffles.cuh'
+] %}
+
+package:
+  name: faiss
+  version: {{ version }}
+
+source:
+  url: https://github.com/facebookresearch/faiss/archive/v{{ version }}.tar.gz
+  sha256: e1a41c159f0b896975fbb133e0240a233af5c9286c09a28fde6aefff5336e542
+  patches:
+    - patches/0001-change-python-Makefile-to-point-to-libfaiss.-SHLIB_E.patch
+    - patches/0002-remove-fPIC-flag-for-msvc-compat.patch  # [win]
+
+build:
+  number: 0
+  # TODO: figure out compiler errors on windows
+  skip: true                      # [win]
+  # TODO: remove this to retry building GPU on the feedstock
+  skip: true                      # [cuda_compiler_version != "None"]
+  # GPU version only for linux; see also conda_build_config.yaml
+  skip: true                      # [(osx or win) and cuda_compiler_version != "None"]
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+
+outputs:
+  - name: _faiss-proc
+    version: 1.0.0
+    build:
+      number: 0
+      string: {{ faiss_proc_type }}
+    test:
+      commands:
+        - exit 0
+    about:
+      home: https://github.com/conda-forge/faiss-feedstock
+      license: MIT
+      license_family: MIT
+      summary: A meta-package to select CPU or GPU build for faiss.
+
+  - name: libfaiss-{{ faiss_proc_type }}
```

Comment:
Remove the `-cpu` here?

Reply:
The names can be bikeshedded of course, but I think a good argument for what I implemented is that the existing packages in they pytorch channel are already named `faiss-cpu` and `faiss-gpu` (so there's no need for a further mutex to enable migrating to the cf-builds etc.).

I also happen to like being explicit in the naming so the installed variant is immediately apparent from `conda-list`, and seeing `faiss-cpu` also makes it quite clear that `faiss-gpu` might exist, whereas just `faiss` would not provide that clue. 

Some more [notes](https://github.com/conda-forge/staged-recipes/pull/11337#issuecomment-625762828) here.

Reply:
You can have the variant name `cpu` or `gpu` in the build string  as in https://github.com/conda-forge/arrow-cpp-feedstock/blob/0.16.x/recipe/meta.yaml#L46

Problem with two names is that the way that it is currently setup, `faiss-cpu` and `faiss-gpu` can both be installed into the same environment. changing the name to `faiss` for both prevents that.

Reply:
I know about build strings, but this was dropped from `ucx-split` for example, with an explanation that it's harder to maintain see this [commit](https://github.com/conda-forge/ucx-split-feedstock/commit/318ee988e899b6e74cc1c03146bcd5728b65e2e1#diff-e178b687b10a71a3348107ae3154e44c). @jakirkham?

> @isuruf: Problem with two names is that the way that it is currently setup `faiss-cpu` and `faiss-gpu` can both be installed into the same environment.

I was aiming at preventing exactly that, see `run_exports` / `run_constrained` here. Granted, I may not be using it 100% correctly, but the intent to leverage this mechanism is there.

Reply:
I haven't seen it being hard to maintain at all.

What prevents a user from being able to do `conda create -n test faiss-cpu faiss-gpu`?

Reply:
> I haven't seen it being hard to maintain at all.

I can't speak to that, just saw the comment on the ucx-feedstock. I'd be OK with build strings, but most people don't look at those, so I'd still prefer the package name to reflect it

> What prevents a user from being able to do `conda create -n test faiss-cpu faiss-gpu`?

That this should result in a conflict of the two different `run_constrained` conditions?

Reply:
> That this should result in a conflict of the two different run_constrained conditions?

`run_constrained` doesn't work like that. `run_constrained` applies only if the package `_faiss-proc` is installed which is not going to be installed in the above command.

Reply:
> I'd be OK with build strings, but most people don't look at those, so I'd still prefer the package name to reflect it 

If a package wants to declare a dependency on `faiss` and doesn't care if it is cpu or gpu, how are they going to declare this in `run`?

Reply:
> run_constrained doesn't work like that. run_constrained applies only if the package _faiss-proc is installed which is not going to be installed in the above command.

But isn't that where `run_exports` comes in and forces it to be installed?

> If a package wants to declare a dependency on faiss and doesn't care if it is cpu or gpu, how are they going to declare this in run?

Fair point. I'm not completely against this, as I said.
I'm guessing in that case that we kind of ignore the transition pain for people on `pytorch::faiss-{cpu,gpu}` to `conda-forge::faiss`, or is there a way to mitigate it?

Reply:
> But isn't that where run_exports comes in and forces it to be installed?

`run_exports` applies only to dependencies.

> is there a way to mitigate it?

We can add metapackages `faiss-cpu` and `faiss-gpu`, but that makes it confusing for people.

Reply:
> run_exports applies only to dependencies.

OK, learned something new again.

> We can add metapackages faiss-cpu and faiss-gpu, but that makes it confusing for people.

I like that idea, I have to say. Nothing a comment can't fix, IMO.